rumqttd/architecture.md

```diff
@@ -139,5 +139,94 @@ NOTE: Transition to ConnectionReady will schedule the connection for consumption
   - If ConnectionBusy
     - ConnectionBusy -> ConnectionReady
 
+## v2
+
+### Ontology
+
+Here are a few terms to understand before we move on. These terms are used internally, within the codebase, to organize the files and modules.
+
+#### Core Concepts and Entities
+
+0. Input and Output Queues (`ibufs` and `obufs`)
+
+   The basic foundation of `rumqttd` are two instances of `VecDequeue` from `std::collections`. Everything else just manages these and their interfaces to various other parts of the system. The queues are as follows -
+
+   - `ibufs`: All incoming data to `rumqttd` (represented by a `Packet`) is stored immediately in `ibufs` (which holds a `VecDequeue`)
+   - `obufs`: All outgoing data `rumqttd` (represented by `Notification`) is stored immediately in the `obufs` (which holds a `VecDequeue`).
+
+   `obufs` also has an _in-flight queue_ to track packets that have been sent.
+
+1. Broker
+
+   Top-level entity that handles _everything_, including, but not limited to
+
+   - Config Validation
+   - Configuring and creating the Router
+   - Starting the Router
+   - Maintaining a channel that communicates with the router
+
+   When you start the Broker, the following happens ->
+
+   1. The Router is started
+   2. (on a new thread) Creates the metrics server
+   3. (on a new thread) Starts all `mqttv4` servers
+   4. (on a new thread) Starts all `mqttv5` servers
+   5. (on a new thread) Starts the prometheus listener
+   6. (on the same thread) Starts the ConsoleLink to allow
+
+2. Router
+
+   Entity managed by the broker that controls the flow of data in `rumqttd`. The router is responsible for managing, authorizing, and scheduling the flow of data between components within and connected to `rumqttd`.
+
+   <!-- At any given time, the router's event loop must choose between handling incoming messages or sending outgoing messages. -->
+
+   The router is built on the actor pattern, where it works primarily by _reacting_ to events that happen. As and when an event is added to `ibufs`, the router reacts by checking its current state (for a certain connection or message) and dispatching the appropriate action. This can involve changing other in-memory data structures, or writing the message to `obufs` to make sure it's handled by the appropriate link.
```

Comment:
> The router is built on the actor pattern, where it works primarily by _reacting_ to events that happen

AFAIK, actor patterns ( model ) is little different? 

> As and when an event is added to `ibufs`, [...] handled by the appropriate link.

Would need more thought.discuss more about this sentence.